File: python/src/transtex/formatters/ieee.py
"""IEEE formatter implementation."""
from __future__ import annotations

from typing import List, Optional, Tuple

from ..reference import Reference
from .shared import preferred_locator, split_name_with_initials


def format_ieee(reference: Reference) -> str:
    segments = [
        _author_segment(reference),
        *_title_and_container(reference),
        _volume_issue_segment(reference),
        _pages_segment(reference),
        reference.year or "",
        preferred_locator(reference, prefix_doi="doi: "),
    ]
    sentence = ", ".join(segment for segment in segments if segment)
    if sentence:
        sentence += "."
    return sentence


def _author_segment(reference: Reference) -> str:
    authors = [_ieee_name(author) for author in reference.normalized_authors()]
    if not authors:
        return ""
    if len(authors) == 1:
        return authors[0]
    if len(authors) == 2:
        return " and ".join(authors)
    return ", ".join(authors[:-1]) + ", and " + authors[-1]


def _title_and_container(reference: Reference) -> Tuple[Optional[str], Optional[str]]:
    container = reference.primary_container()
    if reference.title:
        title_segment = f'"{reference.title},"'
        if container:
            return f'{title_segment} {container}', None
        return title_segment, None
    return None, container


def _volume_issue_segment(reference: Reference) -> str:
    if reference.volume:
        segment = f"vol. {reference.volume}"
        if reference.issue:
            segment += f", no. {reference.issue}"
        return segment
    if reference.issue:
        return f"no. {reference.issue}"
    return ""


def _pages_segment(reference: Reference) -> str:
    return f"pp. {reference.pages}" if reference.pages else ""


def _ieee_name(author: str) -> str:
    last, initials = split_name_with_initials(author)
    if not last:
        return author.strip()
    joined_initials = " ".join(initials)
    if joined_initials:
        return f"{joined_initials} {last}"
    return last


__all__ = ["format_ieee"]

File: typescript/src/formatters/ieee.ts
import { Reference } from "../reference";
import { normalizePageRange, preferredLocator, splitNameWithInitials } from "./shared";

const IEEE_MAX_AUTHORS = 6;

export function formatIeee(reference: Reference): string {
    const container = reference.primaryContainer();
    const common: Array<string | undefined> = [
        authorSegment(reference),
        titleSegment(reference),
        container,
        volumeIssueSegment(reference),
        pagesSegment(reference),
        reference.year ?? "",
        preferredLocator(reference, "doi: "),
    ];

    if (reference.journal || reference.volume || reference.issue) {
        const sentence = joinIeeeSegments(common.filter(Boolean) as string[]);
        return sentence ? `${sentence}.` : "";
    }

    const fallback: Array<string | undefined> = [
        authorSegment(reference),
        titleSegment(reference),
        reference.booktitle ?? container ?? reference.publisher ?? "",
        reference.place ?? "",
        reference.publisher && reference.booktitle ? reference.publisher : "",
        pagesSegment(reference),
        reference.year ?? "",
        preferredLocator(reference, "doi: "),
    ];
    const sentence = joinIeeeSegments(fallback.filter(Boolean) as string[]);
    return sentence ? `${sentence}.` : "";
}

function authorSegment(reference: Reference): string {
    const authors = reference.normalizedAuthors().map(ieeeName);
    if (authors.length === 0) return "";
    if (authors.length > IEEE_MAX_AUTHORS) return `${authors[0]} et al.`;
    if (authors.length === 1) return authors[0];
    if (authors.length === 2) return `${authors[0]} and ${authors[1]}`;
    return `${authors.slice(0, -1).join(", ")}, and ${authors[authors.length - 1]}`;
}

function titleSegment(reference: Reference): string {
    if (!reference.title) return "";
    return `"${reference.title},"`;
}

function volumeIssueSegment(reference: Reference): string {
    if (reference.volume) {
        const pieces = [`vol. ${reference.volume}`];
        if (reference.issue) pieces.push(`no. ${reference.issue}`);
        return pieces.join(", ");
    }
    if (reference.issue) return `no. ${reference.issue}`;
    return "";
}

function pagesSegment(reference: Reference): string {
    const pages = normalizePageRange(reference.pages);
    return pages ? `pp. ${pages}` : "";
}

function joinIeeeSegments(parts: string[]): string {
    if (parts.length === 0) return "";
    let sentence = parts[0];
    for (let i = 1; i < parts.length; i++) {
        const previous = sentence;
        const current = parts[i];
        const separator = previous.endsWith(',"') ? " " : ", ";
        sentence = `${previous}${separator}${current}`;
    }
    return sentence;
}

function ieeeName(author: string): string {
    const [last, initials] = splitNameWithInitials(author);
    if (!last) return author.trim();
    const joined = initials.join(" ");
    return joined ? `${joined} ${last}` : last;
}

export const __all__ = ["formatIeee"];
